### ISSUE TITLE
BUG - ctrl_stock logic is incorrect

### ISSUE THREAD

**Issue by Noodleyman** [cubecart/v6#1266]
stumbled across this today when trying to figure out why all of the inventory  in the category view was showing out of stock when stock control was disabled.

"ctrl_stock" has a true/false value. True means we are enforcing stock levels, false means we are not. For category view this is done in catalogue.class.php
`$product['ctrl_stock'] = (!$product['use_stock_level'] || $GLOBALS['config']->get('config', 'basket_out_of_stock_purchase') || ($product['use_stock_level'] && $GLOBALS['catalogue']->getProductStock($product['product_id'], null, true))) ? true : false;`

with "Allow out-of-stock purchases" iin store settings disabled, and "Use Stock Level" at the product level unchecked, ctrl_stock = true, but it should be false. 

The logic is flawed, it should be:
`$product['ctrl_stock'] = ($product['use_stock_level'] && $GLOBALS['catalogue']->getProductStock($product['product_id'], null, true) && !$GLOBALS['config']->get('config', 'basket_out_of_stock_purchase')) ? true : false;`

the issue exists in both cubecart and catalogue class files.

**Comment by briansandall** [comment]
I think the variable is just poorly named - ctrl_stock is used in content.category.php lines 111 and 169 as well as content.homepage.php line 52 to flag an item as available for purchase (i.e. not out of stock) or not. 

Given the way the variable is used in those templates, I assert that the logic is sound as is: if a product doesn't use stock levels or out-of-stock purchases are allowed, or it does use stock and is in stock, then ctrl_stock should be true to allow the item to be purchased.

**Comment by Noodleyman** [comment]
I was under the impression that "ctrl_stock" means "Stock Control", so true would mean stock control is being enforced, and false means stock control is not being enforced.

so with "Use Stock Level" unchecked, it should be set as false. 

This may just be me misunderstanding the purpose of ctrl_stock, although I am not sure.

**Comment by briansandall** [comment]
That would be the logical conclusion given the variable name, but its usage in the code suggests otherwise.

Implementing your suggested changes would break the current templates - neither products that do not use stock levels nor products that are out of stock but allow out of stock purchases would display in the store.

We'll need @abrookbanks to confirm, but having dug through this code quite a bit, I'm fairly certain the variable is simply the victim of an unfortunately confusing name choice. ;)

**Comment by abrookbanks** [comment]
It is badly named. Many are... :( I'll get onto this asap.

**Comment by briansandall** [comment]
Coming up with good variable names can be the trickiest part of coding...

How about `ctrl_add_to_basket` signifying it is a flag for whether the product can be added to the basket or not? Or perhaps `ctrl_allow_purchase`?

**Comment by Noodleyman** [comment]
Ah, that explains that then... doh. Misunderstanding on my part. I wouldn't actually recommend renaming it since it's been around a while and will be used by third party skins. if it gets changed they would all require updating.

In an ideal world simply having this documented would be enough, so I think this would be my suggestion. Developer documentation ;)

**Comment by Noodleyman** [comment]
Hmm. there is still a bug here, I may just be having one of those moments where I can't get my head out of a specific thought so would appreciate another opinion. 

It relates to this:
https://github.com/cubecart/v6/issues/1103

using what I thought was correct logic, (now I know otherwise) I added a work around, however that is now invalid. I think the root cause is that the "ctrl_stock" value is becoming true, when it should be false in some cases.

to be specific:
Product A is configued to "Use Stock Levels".
Product A has a stock level of -5
Store setting "Allow out-of-stock purchases" is turned off.

The expectation is that product A is unavailable for purchase and should have "crl_stock" = false, however it is evaluating to true. 

After doing some more testing, the issue is only aparent when you have a negative stock amount. If it is 0, then the checks treat 0 as "false", and it evaluates the ctrl_stock to be false.

It's a simple fix. Simply checking if use_stock_level is 1 and stock_level > 0.
`$product['ctrl_stock'] = (!$product['use_stock_level'] || $GLOBALS['config']->get('config', 'basket_out_of_stock_purchase') || ($product['use_stock_level'] && $GLOBALS['catalogue']->getProductStock($product['product_id'], null, true) > 0)) ? true : false;`

The same logic is used in catalogue and cubecart class files, the change is required in both.

**Comment by briansandall** [comment]
Nice catch - a non-zero value definitely evaluates to true whether positive or negative.

And good call on simply adding a little documentation to avoid unnecessary 3rd party changes. Something along the lines of:
```
// @var ctrl_stock True when a product is considered 'in stock' for purposes of allowing a purchase, either by actually being in stock or via certain settings
```

**Comment by briansandall** [comment]
Zero value bug fixed by PR #1406 . Suggest leaving this issue open until documentation detailing expected behavior of `ctrl_stock` is added at appropriate locations.
